Upgrade to drupal-code-generator v3

Chi-teck opened the pull request:
Drush currently uses DCG 2 to generate code.

### Remaining tasks
- [x] Update `drush generate` command to use PHP attributes instead of annotations
- [x] Update `drush:command-file` generator
- [x] Figure out a way to register third-party generators
- [x] Update Drush tests
- [ ] Implement completion for generator names

The big question is how to deliver this change. Here are a couple complications.
1. DCG 3 is not API compatible with DCG 2. 
2. DCG 3 requires PHP 8.1 and Drupal 10.

Basically there are two options.
1. Wait for a new major Drush release that would only support Drupal 10.
2. Make `drush generate` compatible with both DCG versions.

Chi-teck committed:
Update DCG package

Chi-teck committed:
Update 'GenerateCommands'

Chi-teck committed:
Remove 'GeneratorClassResolver'

Chi-teck committed:
Remove 'GeneratorClassResolver'

Chi-teck committed:
Update Migration generator

Chi-teck committed:
Update DrushAliasFile generator

Chi-teck committed:
Register Drush generators and hide completion option

Chi-teck commented:
This kind of critical, because Drupal 10 has been already released and code generated by DCG 2 is not fully compatible with it. In order to support multiple Drupal versions DCG would have to generate code with deprecations .

weitzman commented:
I dont feel strongly about it, but I think 1. is the best way forward. We could branch for Drush 12 now, with a target release date just before Drupalcon Pittsburgh (early June). Drush 12 would be Drupal 10 minimum. Here is the todo list for Drush 12 https://github.com/drush-ops/drush/milestone/7

Chi-teck committed:
Use PHP attributes instead of annotations for generate command

weitzman committed:
Merge branch '12.x' into dcg-3

weitzman commented:
_Update: ignore below. I had merged composer changes into 11.x by accident._

I changed the base of this PR to 12.x It says it has no merge conflicts but I dont see how composer.lock could not conflict. Please reroll if needed when you get a chance.

Chi-teck committed:
Update composer configuration

Chi-teck committed:
Merge 12.x into dcg-3

Chi-teck commented:
> Figure out a way to register third-party generators

Modules can subscribe to generator info event and register their generators. I am not sure if this is enough to resolve this task.

weitzman commented on src/Commands/generate/GenerateCommands.php:
```diff
@@ -20,40 +25,42 @@ class GenerateCommands extends DrushCommands implements AutoloaderAwareInterface
+        $application->setAutoExit(FALSE);
+
+        $application->addCommands([
+            new MigrationGenerator(),
```
Any interest in moving Migration generator in DCG? If not thats , OK.

Chi-teck replied:
DCG already have a generator for MigrationSource plugins. We could add a simple migration generator under Yaml namespace.

Chi-teck replied:
https://github.com/Chi-teck/drupal-code-generator/commit/6347e6ffea446194e37a8fb2db9067d29b0bd676

weitzman reviewed (commented):
How does a global generator subscribe to the event? Here is what we say about authoring global generators today https://www.drush.org/11.x/generators/#global-generators

Chi-teck commented:
> How does a global generator subscribe to the event? Here is what we say about authoring global generators today https://www.drush.org/11.x/generators/#global-generators

Those need to be registered in the generate command before `$application->run();`

weitzman commented:
I updated the dcf generator a bit in https://github.com/drush-ops/drush/pull/5442. Still needs port to DCG3.

weitzman commented:
For global generators, is there any reason we can't keep using the code in `discoverPsr4Generators` - https://github.com/drush-ops/drush/pull/5371/files#diff-14a8581695da2d03094b84103528a480af8e2891e0590de222aa9b04a8d4d7f2L137-L144. That can be called right from the command method.

Chi-teck committed:
Merge branch 12.x into dcg-3

Chi-teck committed:
Update DrushCommandFile generator

Chi-teck committed:
Remove migration generator

Chi-teck commented:
> I updated the dcf generator a bit in #5442. Still needs port to DCG3.

The dcf command has been updated to DCG3. Needs review.

Chi-teck commented:
> For global generators, is there any reason we can't keep using the code in `discoverPsr4Generators` - https://github.com/drush-ops/drush/pull/5371/files#diff-14a8581695da2d03094b84103528a480af8e2891e0590de222aa9b04a8d4d7f2L137-L144. That can be called right from the command method.

The link seems broken. Anyway we can use any discovery method as long as it can ignore DCG 2 generators. I would prefer something less magic, something that does not rely on reflection API and loading generator classes. That would allow to avoid difficulties in future major updates of DCG and symfony/console.

weitzman commented:
The diff wont show because it is inside the file thats deleted in the PR.

IMO we should merge this without major changes to generator discovery in drush (both global and module). Then we work on those in #4685

weitzman committed:
PHPCS

weitzman committed:
Merge branch '12.x' into dcg-3

Chi-teck committed:
Update \Unish\CreateEntityType

Chi-teck committed:
Verbose output in tests

Chi-teck committed:
Remove value for the verbose option

Chi-teck committed:
Revert verbose option

Chi-teck committed:
update CustomGenerator to DCG 3

Chi-teck committed:
Clean-up drush command file generator

Chi-teck committed:
Fix code style issues

weitzman committed:
Add back generator discovery

weitzman committed:
PHPCS

weitzman committed:
PHPCS2

Chi-teck committed:
Adjust generators

Chi-teck committed:
Clean-up

Chi-teck committed:
Remove redundant condition

Chi-teck committed:
Fix code style issue

Chi-teck commented:
I've removed `discoverDrushGenerators`. I don't think it's worth implementing a discovery method for just a couple of generators. Also replaced `adjustDCGGenerators` with event subscriber.

weitzman committed:
Generate only works with bootstrapped Drupal

weitzman committed:
Merge branch '12.x' into dcg-3
